pkg/logbridge/logbridge.go

```diff
@@ -0,0 +1,116 @@
+package logbridge
+
+import (
+	"bufio"
+	"io"
+	"time"
+
+	"github.com/square/p2/pkg/logging"
+)
+
+// Copy implements a buffered copy operation between dest and src.
+// It returns the number of dropped messages as a result of insufficient
+// capacity
+// The caller is responsible for closing & flushing the reader and writer.
+func Copy(dest io.Writer, src io.Reader, capacity int, logger logging.Logger) {
+	lines := make(chan []byte, capacity)
+	droppedMessages := make(chan []byte, capacity)
+	go func(dest io.Writer, src io.Reader, lines chan []byte) {
+		defer close(lines)
+
+		scanner := bufio.NewScanner(src)
+		var line []byte
+		for scanner.Scan() {
+			line = scanner.Bytes() // consume a line regardless of the state of the writer
+			select {
+			case lines <- line:
+			default:
+				select {
+				case droppedMessages <- line:
+				default:
+				}
+				logger.WithField("dropped line", line).Errorln("Dropped was dropped due to full capacity. If this occurs frequently, consider increasing the capacity of this logbridge.")
+			}
+		}
+		if err := scanner.Err(); err != nil {
+			logger.WithError(err).Errorln("Encountered error while reading from src. Proceeding.")
+		}
+	}(dest, src, lines)
+
+	go func(droppedMessages chan []byte) {
+		defer close(droppedMessages)
+
+		numDropped := 0
+		for range droppedMessages {
```

Comment:
Right, I'm thinking that a `defer close(droppedMessages)` at the top of `Copy(…)` will do the trick.